### PR TITLE
[Wasm GC] Refine results in SignatureRefining

### PR DESCRIPTION
**Pull request by kripken**
Similar to what DeadArgumentElimination does for individual functions, this
can refine the results of a set of functions all using the same heap type, when
they all return something more specific. After this PR SignatureRefining can
refine both params and results and is basically complete.

On j2wasm this helps a little, I see around 100 casts end up removed by
followup opts after this runs. However that is a small fraction of the many
thousands of casts and I don't see a benchmark improvement.

### CODE REVIEW

File: test/lit/passes/signature-refining.wast [hunk 1]
```diff
@@ -490,3 +490,137 @@
     )
   )
 )
+
+(module
+  ;; CHECK:      (type $struct (struct_subtype  data))
+  (type $struct (struct_subtype data))
```
Comment (aheejin) [inline test/lit/passes/signature-refining.wast:496]: Is there a doc that defines these new syntax? (`struct_subtype`, `func_subtype`, ...) It sounds like this `struct` is a subtype of `data`, but this doesn't specify any more fields. Does that mean `struct` is the same as `data`?

File: test/lit/passes/signature-refining.wast [hunk 2]
```diff
@@ -490,3 +490,137 @@
...
+  (type $sig-unreachable (func_subtype (result anyref) func))
+
+  ;; CHECK:      (type $none_=>_none (func_subtype func))
+
+  ;; CHECK:      (elem declare func $func-can-refine)
+
+  ;; CHECK:      (func $func-can-refine (type $sig-can-refine) (result (ref $struct))
+  ;; CHECK-NEXT:  (struct.new_default $struct)
+  ;; CHECK-NEXT: )
+  (func $func-can-refine (type $sig-can-refine) (result anyref)
```
Comment (aheejin) [inline test/lit/passes/signature-refining.wast:518]: In the new nominal type syntax, do we write both `type` and `result` in a function signature? Doesn't `type` include `result`? Why do we need to specify `result` redundantly?

**Review by aheejin — COMMENTED** [review]
LGTM, but I have a few questions mostly about the new wast syntax, which I am new to.

File: src/passes/SignatureRefining.cpp [hunk 3]
```diff
@@ -134,20 +147,55 @@ struct SignatureRefining : public Pass {
         }
         newParamsTypes.push_back(lub.getBestPossible());
       }
+      Type newParams;
       if (newParamsTypes.size() < numParams) {
```
[the remark this answers is not in the document]
Comment (kripken) [reply src/passes/SignatureRefining.cpp:151]: Unless I'm missing something, I think that once DCE runs then all the parameters will have information. So adding code to handle the unreachable case is not that worthwhile.

File: test/lit/passes/signature-refining.wast (hunk 1 above)
Comment (kripken) [reply test/lit/passes/signature-refining.wast:496]: I think this is the doc: https://docs.google.com/document/d/1DklC3qVuOdLHSXB5UXghM_syCh-4cMinQ50ICiXnK3Q/edit#

`(type $struct (struct_subtype data))` means a struct with no fields. Not very interesting but it is a unique nominal type which is enough for some tests.

File: test/lit/passes/signature-refining.wast (hunk 2 above)
Comment (kripken) [reply test/lit/passes/signature-refining.wast:518]: Yes, the type name does imply the types of the params and the results, so there is redundancy here. But two functions with the same params and results can now have different nominal types (that is, the function's types are structurally the same, but not nominally). So we need to write out the type name. We could perhaps not write out the params and results in that case, but I'm not sure the wat format allows such shorthand? Even if it does though it seems clearer to write it out I think (so people don't need to scroll up to find the type's definition).

File: src/passes/SignatureRefining.cpp [hunk 4]
```diff
@@ -32,6 +32,7 @@
 #include "ir/lubs.h"
 #include "ir/module-utils.h"
 #include "ir/type-updating.h"
+#include "ir/utils.h"
```
[the remark this answers is not in the document]
Comment (kripken) [reply src/passes/SignatureRefining.cpp:33]: Thanks, fixed!

**Review by tlively — APPROVED** [review]

File: src/passes/SignatureRefining.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (aheejin) [reply src/passes/SignatureRefining.cpp:149]: Is it possible that some of the 'noted' expressions were not `unreachable` but `RefNull` so the `LUBFinder` was not considered as 'noted'? https://github.com/WebAssembly/binaryen/blob/e5594bbde36ac96b6f2a41259d8a8d66e5d0a7cc/src/ir/lubs.h#L51-L52

[the remark this answers is not in the document]
Comment (kripken) [reply src/passes/SignatureRefining.cpp:149]: Oh, good point! Yes, I'll look into a followup...

**Pull request merged**